proposals/0153-elgamal-proof-program.md

```diff
@@ -0,0 +1,240 @@
+    /// This instruction will be left unchanged.
+    VerifyBatchedRangeProofU128,
+
+    /// Verify 256-bit batched range proof.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyBatchedRangeProofU256,
+
+    /// Verify a ciphertext-commitment equality proof.
+    ///
+    /// A ciphertext-commitment equality proof certifies that an ElGamal
+    /// ciphertext and a Pedersen commitment encrypt/encode the same message.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyCiphertextCommitmentEquality,
+
+    /// Verify a grouped-ciphertext with 2 handles validity proof.
+    ///
+    /// A grouped-ciphertext validity proof certifies that a grouped ElGamal
+    /// ciphertext is well-defined, i.e. the ciphertext can be decrypted by
+    /// private keys associated with its decryption handles.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyGroupedCiphertext2HandlesValidity,
+
+    /// Verify a batched grouped-ciphertext with 2 handles validity proof.
+    ///
+    /// A grouped-ciphertext validity proof certifies that a grouped ElGamal
+    /// ciphertext is well-defined, i.e. the ciphertext can be decrypted by
+    /// private keys associated with its decryption handles.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyBatchedGroupedCiphertext2HandlesValidity,
+
+    /// Verify a fee sigma proof.
+    ///
+    /// A `VerifyFeeSigma` proof certifies that a tuple of Pedersen commitments
+    /// satisfy a percentage relation.
+    ///
+    /// This instruction will be RENAMED to `VerifyPercentageWithCap`.
+    VerifyFeeSigma,
+
+    /// Verify a grouped-ciphertext with 3 handles validity proof.
+    ///
+    /// A grouped-ciphertext validity proof certifies that a grouped ElGamal
+    /// ciphertext is well-defined, i.e. the ciphertext can be decrypted by
+    /// private keys associated with its decryption handles.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyGroupedCiphertext3HandlesValidity,
+
+    /// Verify a batched grouped-ciphertext with 3 handles validity proof.
+    ///
+    /// A grouped-ciphertext validity proof certifies that a grouped ElGamal
+    /// ciphertext is well-defined, i.e. the ciphertext can be decrypted by
+    /// private keys associated with its decryption handles.
+    ///
+    /// This instruction will be left unchanged.
+    VerifyBatchedGroupedCiphertext3HandlesValidity,
+}
+```
+
+The implementation of the instructions that are not removed will remain the
+same.
+
+Instead of the modifying the existing ZK Token Proof program into ZK ElGamal
+Proof program, we propose to simply remove the ZK Token Proof program altogether
+and add a new built-in ZK ElGamal Proof program. The existing ZK Token Proof
+program is not yet activated on any of the clusters.
+
+## Impact
+
+The existing ZK Token Proof program will be deprecated and removed. A new ZK
+ElGamal Proof program will be added to the list of built-in programs.
```

Comment:
Can you add the address of the new proof program?

Reply:
Good idea. Done!